src/ast/FuncDecl.ts

```diff
@@ -2,33 +2,51 @@ import {AstNode} from "./AstNode";
 import {VarList} from "./VarList";
 import CommentDecl from "./CommentDecl";
 import {Tokenizer} from "../util/Tokenizer";
+import StaticDecl from "./StaticDecl";
+import AsyncDecl from "./AsyncDecl";
+import ReturnDecl from "./ReturnDecl";
 
 /**
  * Represents a function declaration in our language
  *
- * e.g. function modifier (async | static)? str
+ * e.g. function modifier static? async? str
  *          params []
  *          comments?
- *          (generate getter)?
- *          (generate setter)?
  *          (return TYPE)?
  */
 export default class FuncDecl extends AstNode {
 
     modifier: string;
-    isAsync: boolean;
-    isStatic: boolean;
+    maybeStatic: StaticDecl;
+    maybeAsync: AsyncDecl;
     name: string;
     params: VarList;
     comment: CommentDecl;
-    generateGetter: boolean;
-    generateSetter: boolean;
-    returnType: string;
+    returnDecl: ReturnDecl;
+
 
     public parse(context: Tokenizer): any {
+        let indentLevel: number = context.getCurrentLineTabLevel();
+        context.getAndCheckNext("function");
+        this.modifier = context.getNext();
+
+        this.maybeStatic = new StaticDecl();
+        this.maybeStatic.parse(context);
+        this.maybeAsync = new AsyncDecl();
+        this.maybeAsync.parse(context);
+
+        this.name = context.getNext();
+        if (context.getCurrentLineTabLevel() <= indentLevel) return;
+
+        context.getAndCheckNext("params");
```

Comment:
realized that in a case like this: 
`    function public getTime
        returns string`
This would fail, since we are checking for "params" and throwing an error if it doesn't exist (but params is optional)